COMPACT_DENYLIST_FORMAT.md

```diff
@@ -0,0 +1,457 @@
+  denylists override rules from earlier denylists on conflict.
+
+### Denylist format
+
+#### Summary
+
+The following example showcases the features and syntax of a compact denylist:
+
+```
+version: 1
+name: IPFSCorp blocking list
+description: A collection of bad things we have found in the universe
+author: abuse-ipfscorp@example.com
+hints:
+  hint: value
+  hint2: value2
+---
+# Blocking by CID - blocks wrapped multihash.
+# Does not block subpaths.
+/ipfs/bafybeihvvulpp4evxj7x7armbqcyg6uezzuig6jp3lktpbovlqfkuqeuoq
+
+# Block all subpaths
+/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*
+
+# Block some subpaths (equivalent rules)
+/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*
+/ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*
+
+# Block some subpaths with exceptions: last-matching-rule wins (!)
+/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked*
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blockednot
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/not
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/exceptions*
+
+# Block IPNS domain name
+/ipns/domain.example
+
+# Block IPNS domain name and path
+/ipns/domain2.example/path
+
+# Block IPNS key - blocks wrapped multihash.
+/ipns/k51qzi5uqu5dhmzyv3zac033i7rl9hkgczxyl81lwoukda2htteop7d3x0y1mf
+
+# Block all mime types with exceptions
+/mime/image/*
+!/mime/image/gif
+
+# Legacy CID double-hash block
+# sha256(bafybeiefwqslmf6zyyrxodaxx4vwqircuxpza5ri45ws3y5a62ypxti42e/)
+# blocks only this CID
+//d9d295bde21f422d471a90f2a37ec53049fdf3e5fa3ee2e8f20e10003da429e7
+
+# Legacy Path double-hash block
+# Blocks bafybeiefwqslmf6zyyrxodaxx4vwqircuxpza5ri45ws3y5a62ypxti42e/path
+# but not any other paths.
+//3f8b9febd851873b3774b937cce126910699ceac56e72e64b866f8e258d09572
+
+# Double hash CID block
+# base58btc-sha256-multihash(QmVTF1yEejXd9iMgoRTFDxBv7HAz9kuZcQNBzHrceuK9HR)
+# Blocks bafybeidjwik6im54nrpfg7osdvmx7zojl5oaxqel5cmsz46iuelwf5acja
+# and QmVTF1yEejXd9iMgoRTFDxBv7HAz9kuZcQNBzHrceuK9HR etc. by multihash
+//QmX9dhRcQcKUw3Ws8485T5a9dtjrSCQaUAHnG4iK9i4ceM
+
+# Double hash Path block using blake3 hashing
+# base58btc-blake3-multihash(gW7Nhu4HrfDtphEivm3Z9NNE7gpdh5Tga8g6JNZc1S8E47/path)
+# Blocks /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
+# /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
+# /ipfs/f01701e20903cf61d46521b05f926ba1634628d0bba8a7ffb5b6d5a3ca310682ca63b5ef0/path etc...
+# But not /path2
+//QmbK7LDv5NNBvYQzNfm2eED17SNLt1yNMapcUhSuNLgkqz
+```
+
+#### High level list format
+
+A denylist is made of an optional header and a list of blockitems separated by newlines. Comment lines start with `#`. Empty lines are allowed.
+
+```
+<header>
+---
+<block_item1> [hint_list]
+
+# comment
+<block_item2> [hint_list]
+...
+```
+
+#### Header
+
+The list header is a YAML block:
+
+- Must be valid YAML
+- Fully optional
+- 1KB maximum size
```

Comment:
This seems very short. A description alone could run over. What is the expected processing behaviour in case the size is exceeded?

Reply:
We can increase the max size to whatever is sensible.

If the denylist does not comply with the spec, then you can consider it invalid and stop processing it as a general rule.

Reply:
The reason I'm asking all these annoying questions is because I think that this is a spec for which interoperability is vital. If it were just a convenient way of sharing some configuration or something like an adblock list where the worst-case failure is that you see an ad you don't want to see, then the details of implementation behaviour wouldn't matter much. But the failure impact here is that people could end up going to jail, being fined, or having their business blocked, put on the SafeBrowsing list, etc.. As an operator, if I subscribe to a CSAM or DMCA or phishing blocklist, I need to _know_ that it works.

With 10^2 implementations running over 10^6-10^7 nodes (and more!), knowing that it works means we have to be painfully finicky about processing rules. If your implementation reads 1000 bytes before giving up looking for `---` and mine reads 1024, we might get different lists. Or mine assumes UTF-8 (there's no specification of encoding for the format) but you're Japanese and so used UTF-16, we might get wildly different results, or might be counting characters and bytes differently, etc.

I think you specified 1kb here because that's how far you scan for `---`? I think that that's likely to cause problems over time (more metadata gets added leading to needing more space, but can it interoperate with deployed code, version indicators work poorly in distributed environments, etc.). Might an approach like Markdown frontmatter (having a `---` both before and after) not work better?

More generally, would you be okay if I PRed your PR to make it tighter from an interop perspective? It'll be annoying, and it'll make it longer and less pleasant to read, but I'm afraid that we need it for this context.

Reply:
> I need to _know_ that it works.

The spec cannot ensure this. The implementation may end up failing even if spec is 100% tight on details.



I agree it is good to be more accurate though. What is a reasonable header size?

UTF-8 encoding should also be specified.



> Might an approach like Markdown frontmatter (having a `---` both before and after) not work better?

I don't quite see the point as it doesn't make anything easier with regards to parsing. If `---` is found at the beginning, then you can assume that here is a header but then have to deal with parsing that header. There might still not be a header at all if the denylist is malformed, so you will still have to handle that. You still need to scan until the `---` at the end is found or hit the limit etc.

It is simpler imho the way it is.

It is more interesting to discuss about failure modes: should a malformed entry abort the full list? be ignored? what happens if a header is not found within the limits but the first entry is not a parsable blocking rule (i.e. could signal that the header is too long).

For the the implementation should attempt to block what it can understand and skip over what it cannot understand (logging errors as needed). We should not be blocking or breaking the application due to malformed lists.

Reply:
I have clarified/added details on how to parse the header. I have increased the max size to 1MiB and specified exactly the number of bytes. I have made explicit that the list is UTF-8 encoded and newlines are `\n` characters.